rl4co/envs/scheduling/jssp.py

```diff
@@ -0,0 +1,665 @@
+            device=self.device,
+        )
+        self.durations_cp = torch.zeros(
+            (self.num_jobs, self.num_machines),
+            dtype=torch.float32,
+            device=self.device,
+        )
+
+    def done(self):
+        if len(self.partial_sol_sequence) == self.num_tasks:
+            return torch.tensor(True)
+        return torch.tensor(False)
+
+    def _step(self, td: TensorDict) -> TensorDict:
+        job_idx = td["action"].squeeze()
+        # action is an int 0 - num_jobs
+        # the actual action is the operation id of the selected job
+        # taken from the feasible actions:
+        action = self.feasible_actions[job_idx].clone()  # range 0 - num_tasks
+
+        if action not in self.partial_sol_sequence:
+            # UPDATE BASIC INFO:
+            row = action // self.num_machines
+            col = action % self.num_machines
+            self.finished_mark[row, col] = 1
+            op_duration = self.durations[row, col]
+            self.partial_sol_sequence.append(action)
+
+            # UPDATE STATE:
+            # permissible left shift
+            start_time, flag = permissible_left_shift(
+                action=action,
+                durations=self.durations,
+                machines=self.machines,
+                machines_start_times=self.machines_start_times,
+                operations_on_machines=self.operations_on_machines,
+            )
+            self.flags.append(flag)
+            # update omega or mask
+            if action not in self.last_col:
+                self.feasible_actions[action // self.num_machines] += 1
+            else:
+                self.mask[action // self.num_machines] = 1
+
+            self.ending_times[row, col] = start_time + op_duration
+
+            self.LBs = end_time_lb(self.ending_times, self.durations_cp)
+
+            # adj matrix
+            precd, succd = get_action_nbghs(action, self.operations_on_machines)
+            self.adjacency[action] = 0
+            self.adjacency[action, action] = 1
+            if action not in self.first_col:
+                self.adjacency[action, action - 1] = 1
+            self.adjacency[action, precd] = 1
+            self.adjacency[succd, action] = 1
+            if (
+                flag and precd != action and succd != action
+            ):  # Remove the old arc when a new operation inserts between two operations
+                self.adjacency[succd, precd] = 0
+
+        # prepare for return
+        features = torch.concatenate(
+            (
+                self.LBs.reshape(-1, 1) / configs.et_normalize_coef,
+                self.finished_mark.reshape(-1, 1),
+            ),
+            dim=1,
+        )
+        reward = -(self.LBs.max() - self.max_end_time)
+        if reward == 0:
+            reward = torch.tensor(configs.rewardscale)
+            self.positive_reward += reward
+        self.max_end_time = self.LBs.max()
+
+        tensordict = TensorDict(
```

Comment:
[Minor] in routing problems (e.g. [here](https://github.com/ai4co/rl4co/blob/34fe7a38080b34f0ebb643bc5f7a6fde427d065d/rl4co/envs/routing/cvrp.py#L108)), we simply use `td.update(...)` here which is slightly faster, since we don't need to re-write variables that don't change over time. 

---

NB: we make this with a "trick" in the base env, [this ](https://github.com/ai4co/rl4co/blob/34fe7a38080b34f0ebb643bc5f7a6fde427d065d/rl4co/envs/common/base.py#L96)might be useful in case it seems not to work